Public/Src/Cache/ContentStore/Distributed/NuCache/ContentLocationDatabase.cs

```diff
@@ -665,42 +677,52 @@ public void ContentTouched(OperationContext context, ShortHash hash, UnixTime ac
-                writer.WriteCompact(entry.ContentSize);
-                entry.Locations.Serialize(writer);
-                writer.Write(entry.CreationTimeUtc);
-                long lastAccessTimeOffset = entry.LastAccessTimeUtc.Value - entry.CreationTimeUtc.Value;
-                writer.WriteCompact(lastAccessTimeOffset);
+                serializeFunc(instance, writer);
                 return pooledWriter.Instance.Buffer.ToArray();
             }
         }
 
         /// <summary>
-        /// Deserialize <see cref="ContentLocationEntry"/> from an array of bytes.
+        /// Uses an object pool to fetch a binary reader and feed it into the deserialization function.
         /// </summary>
-        protected ContentLocationEntry Deserialize(byte[] bytes)
+        /// <remarks>
+        /// Be mindful of avoiding lambda capture when using this function.
+        /// </remarks>
+        protected T DeserializeWithPool<T>(byte[] bytes, Func<BuildXLReader, T> deserializeFunc)
         {
             using (PooledObjectWrapper<StreamBinaryReader> pooledReader = _readerPool.GetInstance())
             {
                 var reader = pooledReader.Instance;
-                return reader.Deserialize(new ArraySegment<byte>(bytes), r =>
-                                                                         {
-                                                                             var size = r.ReadInt64Compact();
-                                                                             var locations = MachineIdSet.Deserialize(r);
-                                                                             var creationTimeUtc = r.ReadUnixTime();
-                                                                             var lastAccessTimeOffset = r.ReadInt64Compact();
-                                                                             var lastAccessTime = new UnixTime(creationTimeUtc.Value + lastAccessTimeOffset);
-                                                                             return ContentLocationEntry.Create(locations, size, lastAccessTime, creationTimeUtc);
-                                                                         });
+                return reader.Deserialize(new ArraySegment<byte>(bytes), deserializeFunc);
             }
         }
 
+        /// <summary>
+        /// Serialize a given <paramref name="entry"/> into a byte stream.
+        /// </summary>
+        protected byte[] SerializeContentLocationEntry(ContentLocationEntry entry)
+        {
+            return SerializeWithPool(entry, (instance, writer) => instance.Serialize(writer));
+        }
+
+        /// <summary>
+        /// Deserialize <see cref="ContentLocationEntry"/> from an array of bytes.
+        /// </summary>
+        protected ContentLocationEntry DeserializeContentLocationEntry(byte[] bytes)
+        {
+            return DeserializeWithPool(bytes, ContentLocationEntry.Deserialize);

```

Comment:
If you are series about your comment at line 699, then you should change this method because it allocates, at least once, but maybe more than once.

A bit of C# trivia:

Action a = MyClass.FooBar; // always allocates, this is called method group conversion

Action b = () => MyClass.FooBar(); // won't allocate because C# compiler will cache the delegate.